assets/mdk/brittleness.mdk

```diff
@@ -0,0 +1,179 @@
+
+Consider the following module declaring a `TriangleSum` function and assuming some lemmas about its behavior (which happen to be sufficient to specify it precisely).
+
+<!-- inline-dafny TriangleSum/A1 -->
+
+Based on these lemmas, we can prove that the result of `TriangleSum` is equivalent to a closed formula: `(n * (n + 1)) / 2`. This proof is naturally an inductive one, but Dafny doesn’t automatically know to apply induction because it doesn’t see a recursive function definition. So we need to do at least a little work to tell it to depend on a proof of the identity for `n - 1`  when proving the identity for any `n` bigger than zero. And we need to tell Dafny to use the two provided lemmas. Applying these two ideas leads to the following lemma.
+
+<!-- inline-dafny TriangleSum/A2 -->
+
+Dafny is able to prove this, but it takes over 1M RU to do so. This is partly because the postconditions from both lemmas are in scope everywhere, even though one is useful only for the base case and one is useful only for the inductive case. We can tell Dafny when to use each lemma by including a conditional statement.
+
+<!-- inline-dafny TriangleSum/A3 -->
+
+When conditioning the proof on whether `n` is zero or not, we can invoke the base case lemma only when it is, and invoke the inductive case lemma only when it is not. In addition, we use `assert by` to specialize the result of `TriangleSumRec` to the specific instantiation we need for the proof. With these two additional bits of structure, the resource use goes down to around 100K.
+
+Note that invoking the assumed lemmas about `TriangleSum` in the correct place is critical. If we use the same conditional structure but invoke both at the beginning, so they scope over the entire lemma, Dafny now uses over 1M RU again.
+
+<!-- inline-dafny TriangleSum/A4 -->
+
+# More general guidelines
+
+The examples covered so far illustrate two key principles for maintainable verification:
+
+* Break long leaps of reasoning down into smaller steps. This limits the number of reasoning steps that the prover needs to construct on its own.
+* Isolate each step so that the proof of it doesn’t need to (and, further, isn’t able to) take into account irrelevant information. This limits the number of choices that the prover has to evaluate when making each reasoning step.
+
+A nice coincidence is that programs written according to these principles tend to be more understandable to humans, as well. Reducing the work a prover needs to do when reasoning about a program has the effect of reducing the amount of thinking a human has to do to understand what a program does, and why it’s correct. Constructing small, relatively isolated components relates strongly to the concepts of [cohesion](https://en.wikipedia.org/wiki/Cohesion_(computer_science)) and [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming)) used in software engineering.
```

Comment:
Rather than talking about "understandable to humans" I would just say something like "easier to read"

Reply:
Should we mention somewhere that the rules for structuring code are few and well know? Should mention that they are called natural deduction?